Description: Stop periodic tasks during shutdown
 The oslo.service periodic tasks runner expects that tasks are returning
 a number of seconds indicating when to run next time. During graceful
 shutdown we want to stop the periodic tasks but we returned None which
 breaks the protocol and triggers a stack trace. Instead this patch makes
 sure we raise LoopingCallDone to signal that no more tasks needs to be
 run during shutdown.
Author: Balazs Gibizer <gibi@redhat.com>
Date: Wed, 16 Sep 2026 15:33:31 +0200
Bug: https://launchpad.net/bugs/2167481
Change-Id: I44999b3354ae32f4cdf356b46d60d42e004e2ccc
Signed-off-by: Balazs Gibizer <gibi@redhat.com>
Origin: upstream, https://review.opendev.org/c/openstack/nova/+/1005880
Last-Update: 2026-09-20

Index: nova/nova/manager.py
===================================================================
--- nova.orig/nova/manager.py
+++ nova/nova/manager.py
@@ -55,6 +55,7 @@ import threading
 import time
 
 from oslo_log import log as logging
+from oslo_service import loopingcall
 from oslo_service import periodic_task
 
 import nova.conf
@@ -128,7 +129,11 @@ class Manager(PeriodicTasks, metaclass=M
         """Tasks to be run at a periodic interval."""
         if self._shutdown_in_progress.is_set():
             LOG.debug('Skipping periodic tasks during graceful shutdown.')
-            return
+            # We either need to return a value or raise LoopingCallDone
+            # to avoid a RuntimeError in DynamicLoopingCall. As we
+            # are shutting down, we don't want to run any periodic tasks any
+            # more, so we raise.
+            raise loopingcall.LoopingCallDone()
         return self.run_periodic_tasks(context, raise_on_error=raise_on_error)
 
     def init_host(self, service_ref):
Index: nova/nova/tests/unit/test_manager.py
===================================================================
--- nova.orig/nova/tests/unit/test_manager.py
+++ nova/nova/tests/unit/test_manager.py
@@ -16,6 +16,8 @@ import threading
 import time
 from unittest import mock
 
+from oslo_service import loopingcall
+
 from nova import manager
 from nova import test
 
@@ -55,11 +57,24 @@ class ManagerTaskTrackingTestCase(test.N
     @mock.patch.object(manager.Manager, 'run_periodic_tasks')
     def test_periodic_tasks_skipped_when_shutdown_in_progress(
             self, mock_run):
-        # Periodic tasks are not run when shutdown is in progress
+        # Periodic tasks are not run when shutdown is in progress. We raise
+        # LoopingCallDone to stop the DynamicLoopingCall cleanly. Returning
+        # None would break the protocol and trigger a RuntimeError in the
+        # looping call runner, see bug 2167481.
         self.manager._shutdown_in_progress.set()
-        result = self.manager.periodic_tasks(mock.sentinel.context)
+
+        self.assertRaises(
+            loopingcall.LoopingCallDone,
+            self.manager.periodic_tasks, mock.sentinel.context)
         mock_run.assert_not_called()
-        self.assertIsNone(result)
+
+        call = loopingcall.DynamicLoopingCall(
+            self.manager.periodic_tasks, mock.sentinel.context)
+        # periodic_interval_max=None mirrors our default config
+        call.start(periodic_interval_max=None)
+        # LoopingCallDone's default retvalue is True, so wait() returns True
+        # without raising RuntimeError, see bug 2167481.
+        self.assertTrue(call.wait())
 
         # Periodic tasks runs normally when shutdown is not in progress
         self.manager._shutdown_in_progress.clear()
